lisp/ess-julia.el

```diff
@@ -364,6 +364,52 @@ It makes underscores and dots word constituent chars.")
 
 (defvar ess-julia-mode-syntax-table (copy-syntax-table julia-mode-syntax-table))
 
+(defvar ess-julia-function-pattern
+  ;; (rx (or "function"
+  ;;         "macro"
+  ;;         (seq (0+ nonl) "(" (1+ nonl) ")" (0+ whitespace) "=")))
+  "function\\|macro\\|.*(.+)[[:space:]]*="
+  "Regular expression to match the beginning of a function in Julia buffers.")
+
+(cl-defmethod ess-beginning-of-function--override
+  (no-error &context ((string= ess-dialect "julia") (eql t)))
+  "Find the beginning of the function in Julia."
+  (let ((p (point))
+        (bound (or (julia-last-open-block (point-min))
+                   (point-min))))
+    (goto-char p)
+    ;; Calling `beginning-of-line' here excludes nested fu
+    (beginning-of-line)
+    (while (and (not (looking-at-p ess-function-pattern))
+                (< bound (point)))
+      (forward-line -1)
+      (beginning-of-line))
+    (if (looking-at-p ess-function-pattern)
+        (point)
+      (unless no-error (user-error "Point not inside function")))))
```

Comment:
if the point is outside a function, it will still jump back to the previous function here. Personally, I would throw a real error so it could be caught in a condition-case block